### PR TITLE
Improvements to Container building

### PR DESCRIPTION
**Pull request by badlop**
Here are a lot of small improvements, some fixes and many cosmetic changes, to the main Dockerfile, the Dockerfile.binary-install, and other related files.

Each commit includes only one change, and has a self-explanatory commit message, with additional description when necessary.

### CODE REVIEW

File: .github/container/Dockerfile.binary-install [hunk 1]
```diff
@@ -39,24 +34,28 @@ RUN set -x && echo -e \
...
+            -nodes \
+            -newkey rsa:4096 \
+            -keyout $PEM \
+            -out $PEM \
+            -days 3650 \
+            -subj "/CN=localhost" \
+    && chown 9000:9000 $PEM
+
+RUN apk del \
+        openssl
```
Comment (sando38) [inline .github/container/Dockerfile.binary-install:52]: Shall we also remove the libcap-utils package as it is not of any use after setting the caps for beam.smp? The libraries are part of libcap2 package which will remain in the container

File: .github/container/Dockerfile.binary-install [hunk 2]
```diff
@@ -1,34 +1,29 @@
-FROM alpine:3.17.0 AS ejabberd
-# BUILD: define image build arguments
+FROM alpine:3.17 AS ejabberd
```
Comment (sando38) [inline .github/container/Dockerfile.binary-install:1]: Shall we switch to alpine 3.16 for now to stick to the binary built erlang/otp version 24?

File: .github/container/Dockerfile.binary-install [hunk 3]
```diff
@@ -1,34 +1,29 @@
...
-RUN set -x && apk -U upgrade --available \
-# BUILD: add build dependencies
-    && apk add -t build-deps \
-        libcap-utils \
-# RUNTIME: add/ define runtime packages
+RUN apk -U upgrade --available \
     && apk add \
-        $ADD_CAPTCHA_PKGS \
         busybox-binsh \
+        libcap-utils \
```
Comment (sando38) [inline .github/container/Dockerfile.binary-install:10]: if we move to alpine 3.16 we need to rename this to package name `libcap` see https://pkgs.alpinelinux.org/package/v3.16/main/x86_64/libcap 

if we use alpine 3.17 the added packages `libcap2` and `libcap-utils` are okay!

File: .github/container/Dockerfile.binary-install [hunk 4]
```diff
@@ -39,24 +34,28 @@ RUN set -x && echo -e \
         '#!/bin/sh \
         \nexec "$@"' > /usr/local/bin/run.sh \
     && chmod +x /usr/local/bin/* \
-# RUNTIME: set CAPS for binding to privileged ports & remove unnecessary binaries
+    # RUNTIME: set CAPS for binding to privileged ports & remove unnecessary binaries
```
Comment (sando38) [inline .github/container/Dockerfile.binary-install:37]: the comment is not 100% correct, we do not remove the "unnecessary binaries" :)

**Comment by sando38** [comment]
hi @badlop looks good! I have put in some comments which we can resolve quickly from my point of view!

File: .github/container/Dockerfile.binary-install (hunk 2 above)
Comment (sando38) [reply .github/container/Dockerfile.binary-install:1]: However, we could also bump the `make-binaries` script to use erlang/OTP v25:

```
diff --git a/tools/make-binaries b/tools/make-binaries
index 0eb10a640..84e4c0ef7 100755
--- a/tools/make-binaries
+++ b/tools/make-binaries
@@ -70,8 +70,8 @@ termcap_vsn='1.3.1'
 expat_vsn='2.4.9'
 zlib_vsn='1.2.13'
 yaml_vsn='0.2.5'
-ssl_vsn='1.1.1q'
-otp_vsn='24.3.4.5'
+ssl_vsn='1.1.1s'
+otp_vsn='25.2.1'
 elixir_vsn='1.14.0'
 pam_vsn='1.5.2'
 png_vsn='1.6.38'
@@ -757,7 +757,7 @@ build_rel()
        export CFLAGS="-g0 -O2 -pipe -fomit-frame-pointer -static-libgcc $CPPFLAGS"
        export CXXFLAGS="$CFLAGS -static-libstdc++"
        export LDFLAGS="-L$prefix/lib -static-libgcc -static-libstdc++"
-       export ERL_COMPILER_OPTIONS='[deterministic, no_debug_info]'
+       export ERL_COMPILER_OPTIONS='[no_debug_info]' # Building 25.x fails with 'deterministic'.
 
        if [ "$mode" = 'cross' ]
        then configure="./configure --host=$target --build=$platform"
```

`deterministic` option does not work for some reason, Holger discovered that, but I checked the output and the binaries do not seem to be unusually larger the before.

File: .github/container/Dockerfile [hunk 5]
```diff
@@ -112,7 +112,7 @@ HEALTHCHECK \
 
 WORKDIR $HOME
 USER ejabberd
-VOLUME ["$HOME/conf", "$HOME/database", "$HOME/logs", "$HOME/upload"]
+VOLUME ["$HOME"]
 EXPOSE 1883 4369-4399 5210 5222 5269 5280 5443
 
 ENTRYPOINT ["/usr/local/bin/ejabberdctl"]
```
Comment (sando38) [inline .github/container/Dockerfile:118]: In the Dockerfile.binary-install I included the alpine package `tini` which is a [runtime init](https://github.com/krallin/tini) which improves signal handling of containers and also kills zombie processes. I does not create overhead and is small. So I suggest adding it as well. `ENTRYPOINT` must be adjusted e.g. like this `["/sbin/tini","--","/usr/local/bin/ejabberdctl"]`

**Comment by badlop** [comment]
`libcap-utils` or any other package: if it isn't required at runtime to run ejabberd, or any of its features, or compile and install external modules... then ok, let's delete it/them

`tini`: ok, let's add it to Dockerfile too

`Dockerfile.binary-install` exists as a workaround because QEMU has a problem with OTP 25. There's no reason in using Dockerfile.binary-install with OTP 24 right?

**Comment by sando38** [comment]
> `Dockerfile.binary-install` exists as a workaround because QEMU has a problem with OTP 25. There's no reason in using Dockerfile.binary-install with OTP 24 right?

yeah, exactly. We should avoid from my point of view however that we have different erlang versions:

Currently the binaries are built with erlang/otp v24 and alpine 3.17.0 has erlang/otp v25. This would cause a problem, if we e.g. compile `mod_ecaptcha` with the alpine packages afterwards. 
Easy way would be to update the `make-binaries` script to `25.2.2` so there would be not problem and keep both `make-binaries` and alpine with the same erlang/otp vsn.

**Pull request merged**